lib/pacer.go

```diff
@@ -0,0 +1,170 @@
+package vegeta
+
+import (
+	"fmt"
+	"math"
+	"time"
+)
+
+// A Pacer defines the rate of hits during an Attack by
+// returning the duration an Attacker should wait until hitting
+// the next Target. If the second return value is true, the
+// attack will terminate.
+type Pacer interface {
+	Pace(elapsedTime time.Duration, elapsedHits uint64) (sleep time.Duration, stop bool)
+}
+
+// A PacerFunc is a function adapter type that implements the
+// Pacer interface.
+type PacerFunc func(time.Duration, uint64) (time.Duration, bool)
+
+// Pace implements the Pacer interface.
+func (f PacerFunc) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	return f(elapsedTime, elapsedHits)
+}
+
+// Rate sends a constant rate of hits to the target.
+type Rate struct {
+	Freq int           // Frequency (number of occurrences) per ...
+	Per  time.Duration // Time unit, usually 1s
+}
+
+// Pace determines the length of time to sleep until the next hit is sent.
+func (r Rate) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	if r.Per <= 0 || r.Freq <= 0 {
+		// If Rate configuration is invalid, stop the attack.
+		return 0, true
+	}
+	interval := uint64(r.Per.Nanoseconds() / int64(r.Freq))
+	delta := time.Duration((elapsedHits + 1) * interval)
+	// Zero or negative durations cause time.Sleep to return immediately.
```

Comment:
(opinionated) I would simplify this to zero, avoiding negative cases as they don't make much sense for the caller.

Reply:
This comment is not here for the caller, it is here for the maintainer, to explain why I'm not bothering to check whether `delta - elapsedTime` is negative and clamp to zero :-)

Reply:
What I mean was - I think it would be better to never return negative number from this function. Although `time.Duration` allows for this outcome, I think logically it doesn't make any sense and requires that caller be wary of it ...

If we always return `0` instead of negatives we might help the caller to avoid doing "something stupid". 

Testing it also will be easier (tests are what actually forced me to think about this in the first place, and most of the times testing-first leads to better design ...)

Just trying to simplify things wherever possible :)